compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Check.hs

```diff
@@ -378,101 +380,145 @@ typeOfTmLam (var, typ) body = do
-            CPNil -> typeOf rhs
+            CPNil -> (,) (singletonMR 0) <$> typeOf rhs
             CPCons headVar tailVar
               | headVar == tailVar -> throwWithContext (EClashingPatternVariables headVar)
-              | otherwise -> introExprVar headVar elemType $ introExprVar tailVar (TList elemType) $ typeOf rhs
-            CPDefault -> typeOf rhs
+              | otherwise -> do
+                rhsType <- introExprVar headVar elemType $ introExprVar tailVar (TList elemType) $ typeOf rhs
+                pure (singletonMR 1, rhsType)
+            CPDefault -> (,) AllRanks <$> typeOf rhs
             _ -> throwWithContext (EPatternTypeMismatch patn (TList elemType))
 
-typeOfAltsOptional :: MonadGamma m => Type -> [CaseAlternative] -> m Type
+typeOfAltsOptional :: MonadGamma m => Type -> [CaseAlternative] -> m (MatchedRanks, Type)
 typeOfAltsOptional elemType =
     typeOfAlts $ \(CaseAlternative patn rhs) -> do
         case patn of
-            CPNone -> typeOf rhs
-            CPSome bodyVar -> introExprVar bodyVar elemType $ typeOf rhs
-            CPDefault -> typeOf rhs
+            CPNone -> (,) (singletonMR 0) <$> typeOf rhs
+            CPSome bodyVar -> do
+                rhsType <- introExprVar bodyVar elemType $ typeOf rhs
+                pure (singletonMR 1, rhsType)
+            CPDefault -> (,) AllRanks <$> typeOf rhs
             _ -> throwWithContext (EPatternTypeMismatch patn (TOptional elemType))
 
 -- NOTE(MH): The DAML-LF spec says that `CPDefault` matches _every_ value,
 -- regardless of its type.
-typeOfAltsOnlyDefault :: MonadGamma m => Type -> [CaseAlternative] -> m Type
+typeOfAltsOnlyDefault :: MonadGamma m => Type -> [CaseAlternative] -> m (MatchedRanks, Type)
 typeOfAltsOnlyDefault scrutType =
     typeOfAlts $ \(CaseAlternative patn rhs) -> do
         case patn of
-            CPDefault -> typeOf rhs
+            CPDefault -> (,) AllRanks <$> typeOf rhs
             _ -> throwWithContext (EPatternTypeMismatch patn scrutType)
 
 typeOfCase :: MonadGamma m => Expr -> [CaseAlternative] -> m Type
 typeOfCase scrut alts = do
     scrutType <- typeOf scrut
-    case scrutType of
+    (numRanks, rankToPat, (matchedRanks, rhsType)) <- case scrutType of
         TConApp scrutTCon scrutTArgs -> do
             DefDataType{dataParams, dataCons} <- inWorld (lookupDataType scrutTCon)
             case dataCons of
-                DataVariant cons -> typeOfAltsVariant scrutTCon scrutTArgs (map fst dataParams) cons alts
-                DataEnum cons -> typeOfAltsEnum scrutTCon cons alts
-                DataRecord{} -> typeOfAltsOnlyDefault scrutType alts
-        TUnit -> typeOfAltsUnit alts
-        TBool -> typeOfAltsBool alts
-        TList elemType -> typeOfAltsList elemType alts
-        TOptional elemType -> typeOfAltsOptional elemType alts
-        _ -> typeOfAltsOnlyDefault scrutType alts
+                DataVariant cons -> (,,) (length cons) (\k -> CPVariant scrutTCon (fst (cons !! k)) wildcard)
```

Comment:
It's quite unfortunate that getting the number of constructors is linear in the same number. However, this number is only used if we don't have a default pattern and hence are either linear in the number of constructors already or we'll throw an error and this sort of performance doesn't matter anymore.
Nevertheless, I'll look into using a vector for this field in a separate PR.

Reply:
Yes, that's only used in the error case.